tikv-client/src/main/java/com/pingcap/tikv/expression/visitor/PrunedPartitionBuilder.java

```diff
@@ -0,0 +1,208 @@
+      case EQUAL:
+        ranges.add(Range.singleton(literal));
+        break;
+      case NOT_EQUAL:
+        ranges.add(Range.lessThan(literal));
+        ranges.add(Range.greaterThan(literal));
+        break;
+      default:
+        throwOnError(node);
+    }
+    return ranges;
+  }
+
+  @Override
+  protected RangeSet<Long> visit(LogicalBinaryExpression node, Void context) {
+    RangeSet<Long> leftRanges = node.getLeft().accept(this, context);
+    RangeSet<Long> rightRanges = node.getRight().accept(this, context);
+    switch (node.getCompType()) {
+      case AND:
+        rightRanges.removeAll(leftRanges.complement());
+        break;
+      case OR:
+        rightRanges.addAll(leftRanges);
+        break;
+      case XOR:
+        // AND
+        rightRanges.removeAll(leftRanges.complement());
+        // full set
+        rightRanges.addAll(leftRanges);
+        rightRanges.removeAll(rightRanges);
+        break;
+      default:
+        throwOnError(node);
+    }
+    return rightRanges;
+  }
+
+  public TiPartitionInfo pruning(TiTableInfo tableInfo, List<Expression> filters) {
+    filters = filterFilters(filters);
+
+    Expression cnfExpr = PredicateUtils.mergeCNFExpressions(filters);
+    if (!canBePruned(tableInfo, cnfExpr)) {
+      return tableInfo.getPartitionInfo();
+    }
+
+    List<Expression> partExprs = generatePartExprs(tableInfo);
+    TiPartitionInfo partInfo = tableInfo.getPartitionInfo();
+    RangeSet<Long> filterRange = buildPartRange(cnfExpr);
+    List<TiPartitionDef> pDefs = new ArrayList<>();
+    for (int i = 0; i < partExprs.size(); i++) {
+      Expression partExpr = partExprs.get(i);
+      RangeSet<Long> partRange = buildPartRange(partExpr);
+      partRange.removeAll(filterRange.complement());
+      if (!partRange.isEmpty()) {
+        // part range is empty indicates this partition can be pruned.
+        pDefs.add(partInfo.getDefs().get(i));
+      }
+    }
+    return new TiPartitionInfo(
+        partInfo.getType(), partInfo.getExpr(), partInfo.getColumns(), true, pDefs);
+  }
+
+  private RangeSet<Long> buildPartRange(Expression filter) {
+    return filter.accept(this, null);
+  }
+
+  private static List<Expression> generatePartExprs(TiTableInfo tableInfo) {
+    // only support column ref for now. Fn Expr will be supported later.
+    TiPartitionInfo partInfo = tableInfo.getPartitionInfo();
+    List<Expression> partExprs = new ArrayList<>();
+    for (int i = 0; i < partInfo.getDefs().size(); i++) {
+      TiPartitionDef pDef = partInfo.getDefs().get(i);
+      Constant cst = Constant.create(Longs.tryParse(pDef.getLessThan().get(0)));
+      if (i == 0) {
+        partExprs.add(ComparisonBinaryExpression.lessThan(partExpr, cst));
+      } else {
+        Constant previous =
+            Constant.create(Longs.tryParse(partInfo.getDefs().get(i - 1).getLessThan().get(0)));
+        Constant current =
+            Constant.create(Longs.tryParse(partInfo.getDefs().get(i).getLessThan().get(0)));
```

Comment:
Would you please explain why get(0)? If you are only to deal with range type partition only, please at least leave room for other type in your code. If this method is just to generate expr for range type, name it like generateRangePartExpr. Also guard on partition type at beginning.

Reply:
It is hardcode version and will be replaced by a parser. The parser can handle this well. I should comment on this. Maybe we should first review parser code?